### ISSUE TITLE
Linux Version 

### ISSUE THREAD

**Issue by RinMaru** [project64/project64#669]
Would it be possible to get a Linux Version of Project 64?

**Comment by md500** [comment]
I think is possible. Linux also works in C,C++ but i dont know if libraries are standard...windows.h library  is not standard...

**Comment by Zetchzie** [comment]
It runs in Wine fine but Wine can be slow and buggy from my experience. Just having a Linux version may alleviate these probs.

**Comment by JunielKatarn** [comment]
A lot of code would have to be ported.
I gave it a thought and it could help to port the front-end from C++ to C#. That way, you could have that part of the software running in Windows, Linxu and OS X.

That said, there is probably more Windows-dependent code in the backend functionality and probably in the plugins.

**Comment by cxd4** [comment]
Just in case nobody has been noticing my grand mastermind scheme, most of my changes in the PRs are in fact gradual towards platform-agnosticism.  It has taken only a fraction of the pull requests necessary for me to make up to this point, but my goal is to get it to compile on at least one Unix-based platform.

The next steps currently are to:
* Implement a command-line interface to Project64, now that I have gotten it to build w/o WTL.
* Add a MinGW build script somewhere in the repository to build without using MSVC.

Once those things are done, something like a Linux port should hardly be out of the foreseeable future--it would be down to things like getting rid of custom typedef's by Microsoft and relying more on the CLI.

**Comment by Zetchzie** [comment]
this is good to know ^^

**Comment by md500** [comment]
I think,he must port the emulator to c# and use things as Cmake,Makefiles archives and wxWidgets or MFC library to linux,and consolidate this with alsa,pulseaudio or jack server.in graphics, GLide64 is recomended to linux (GLide64 uses OpenGL i think -but GL 2 is minimun recommendend-) and create plugins for Xrender for the Xorg and Wayland servers but in ubuntu it uses mir server....
Sorry for my english

**Comment by Zetchzie** [comment]
Your english is fine. idk much about mir.

**Comment by cxd4** [comment]
Well I don't want to port if it means having more platform-specific things and #ifdef's all over the place.

My objective behind a Linux port isn't the materialist value of having the thing itself; it's simply to counteract the lockdown of Project64's coding to a single platform running on a silly, arbitrary CPU architecture like x86.  If I were to port Project64 to Mac or Linux or anything else just by adding ALSA, CoreAudio or those other things that you said I'd only be making the code even worse with more combinations of even more platform-restricted code.

So I want it to be simple and work universally.  Command line is one very direct way to achieve this, as I often find CLIs to be easier to use, more direct and faster for me to work with than GUIs anyway.

**Comment by md500** [comment]
Mmm,original Mupen64 uses commanline emulator,but it would better use a CLI menu,such ncurses style.But for Xorg it would better use gtk2 for more compatibity. The thing is what some users dont like commanline users,for configuring plugins,select a rom and other things.so it would use ncurses for CLI,and gtk2 for Xorg server and compatibility for tty's(without Xorg server).but Alsa is more universal,but pulseaudio is important too (some distributions uses it).Better start for CLI menu first for some test and alsa comatibity first.

**Comment by cxd4** [comment]
I like to run Mupen64 with `-nogui` mode.  Only problem is I can't use saved states.

I'm happy if anyone else wants to contribute a graphical alternative in GTK, Qt or whatever, but programmatically it's easiest to extend new options off of CLI.  Not sure what I would need ncurses for.  I think the assumption of shell support from the OS is already enough for the code; I'm not sure what further constrictions the use of ncurses would apply.  Even in the worst case scenario, you could just write a shell script or a batch file that starts playing a ROM in the emulator instantly, which is not something any interactive UIs allow you to do--graphical or otherwise.

ALSA is just for audio isn't it?  The audio plugin can handle that business; currently there are several sound output plugins (now including Azimer's Audio) that can plug in to e.g. the Linux port of any emulator.

That just leaves the windowing framework for drawing graphics--at the bottom of it all you only need X11 for it to work, but many modern distros seem to ship with SDL and FreeGLUT which could also be used for cleaner, higher-level management.  I personally would prefer FreeGLUT.

**Comment by darakian** [comment]
command line seems like a good starting point. I was looking to see how to compile and I don't see a make file. Has anyone tried a compile yet?

**Comment by cxd4** [comment]
As of yet, the only way to compile is with MSVC.
Visual Studio Express is a non-priced way to compile Project64, but we have no GUI.

I can get Project64 to compile in other compilers such as MinGW but not without the resulting build being an interface-less shell without any way to start emulation, so I was thinking that implementing command-line takes the priority first.

**Comment by darakian** [comment]
Ok cool. I guess getting a gcc compile is a lesser priority?

**Comment by cxd4** [comment]
Getting it to compile on any (sane) compiler is a priority.

MinGW is just a Windows rewrite of GCC to port it to Win32.
If it compiles on MinGW, it essentially compiles on GCC.

**Comment by cxd4** [comment]
To be honest I'm not all that anxious to implement a CLI to Project64 either, so just getting it buildable on MinGW or whatever (for the sake of testing compiles without having Windows installed) sounds at least as good to me.  I'm fine with whoever does the UI in gtk, Qt etc..

**Comment by darakian** [comment]
If I can issue 
./project64 some.rom
I'll be very happy. Is there a blog or any sort of central organization for the porting effort?

**Comment by cxd4** [comment]
>./project64 some.rom

That's exactly the goal, but even as I keep contributing more and more portability PRs, I have to decide between a fork in the road, whether to port the GUI to a CLI or to get it to compile on gcc first.

Even right now on Windows, my Visual Studio Express build of Project64 cannot link the C++ ATL GUI that Project64 uses.  You can therefore, on Windows, execute `project64 some.rom` with the VS Express build I have gotten to work, but it will do nothing more than print the argument vector.  Because the current entire UI framework requires Visual Studio Professional with all those proprietary interfaces for WTL.

**Comment by darakian** [comment]
Right on. I for one vote on the gcc compile first (for what it's worth), but again is there a blog or central organization thing? I'd like to help, but I've got no idea where to start.

**Comment by cxd4** [comment]
Sorry nope, just the GitHub discussions, commit logs and arguments here.

The portability pull request I currently have open is this one:  https://github.com/project64/project64/pull/650

It's intended to make Project64 portable to machines with big-endian CPUs (or other byte orders besides x86 32-bit little-endian).  It also gradually replaces some Microsoft types like WORD and DWORD with the standard integer types from C99, which will help both Linux build as well as an Android build.

Oh, but we do have a small IRC channel on Freenode if that's what you meant, but no blog.

**Comment by Zetchzie** [comment]
In the long run I hope the minimal requirements stay the same Project 64 runs at full speed on my  Intel® Pentium(R) 4 CPU 2.80GHz × 2 Intel® 915G x86/MMX/SSE2.

**Comment by darakian** [comment]
That might be a long shot. I was getting lag on a 4770k + 260x earlier today.

**Comment by JunielKatarn** [comment]
Some thoughts here.
Compiler suite:
What about LLVM instead of GCC?
Microsoft is working on clang compiler compatibility and in parallel, LLVM is adding MSVC ABI compatibility. Meaning, you don't need to get rid of MSVC to have the code compile and link in UNIX.

Makefiles:
We already have MsBuild files (vcxproj) which can be configured for UNIX platforms. No need to write new make files from scratch. See xbuild, from the Mono Project (C# for UNIX).

GUI:
I like the CLI idea, but eventually, the majority of users will prefer a graphical interface (mostly non-developers and no-power users). Another approach to keep the GUI and not deal with OS-specifics, is to rewrite it in C# (again, see the http://www.mono-project.com). This way, we could have a cross-platform compile-once user interface, and leave the performance/platform oriented code to the core code (N64 classes) and the plugins.

And yes, C# and xbuild are both free and open source.

This approach would minimize the effort to port to new platforms after Linux (Mac OS X, even BSD).

**Comment by cxd4** [comment]
>What about LLVM instead of GCC?

Clang, LLVM, cc, gnu c/gcc, MinGW even perhaps... all various names for the same end thing really (some of which related to Google's scheming and are less interesting than others).

I don't think making the code work on one should be different than doing so for the others.

> Microsoft is working on clang compiler compatibility and in parallel,

Just to make myself clear, it shouldn't be Clang-specific, it shouldn't be MSVC-specific, and it shouldn't be MSVC-or-Clang-exclusive.  It should just be portable to other compilers.  I don't aim to focus on a given set of compilers, just following standard code rules so that it just works everywhere more often (surely not every compiler in existence, obviously, but as close as reasonably testable).

There may be some contingencies where I may have to install the Clang fork to test breakage, though I anticipate little of that, as it is merely a fork of the same basic compiler.

> LLVM is adding MSVC ABI compatibility.

Frankly, MSVC LLP64 ABI is horrible.  It rebels against every current, sane ABI there is.

With any other ABI, `long` is a 64-bit type when building for a 64-bit target.
Only Windows 64-bit seems to be the exception:  `long` is still 32-bit because Microsoft persisted on backward-compatibility with their 16-bit DOS type rules, where `unsigned long` was required to guarantee a 32-bit DWORD.  (`int` I'm fairly sure would have to have been 16-bit back then.)

So for some reason, maybe LLVM decides to implement this as an optional feature, but using platform- or compiler-specific ABI coding is not the way to go for this, even if one or two other compilers or projects implement compatibility with those platform- or compiler-specific ABIs.  Simple use of even just the C99 standard integer definitions should be more than sufficient for using central type sizes in a way that works on any modern compiler without having to #ifdef between multiple ABIs, compilers or compiler modes or whatever.

>Makefiles:
We already have MsBuild files (vcxproj) which can be configured for UNIX platforms. No need to write new make files from scratch. See xbuild, from the Mono Project (C# for UNIX).

Sounds good to me.  I never really liked GNU makefile syntax anyway.

Makefile wars are not interesting to me; we can stick with whatever MsBuild does like you said.  (Not that I really understand their syntax either, but I can contribute some optional batch file or shell scripts to the repo's sub-sub "Scripts" directory to avoid too much pollution while doing my tests.)

>GUI:
I like the CLI idea, but eventually, the majority of users will prefer a graphical interface (mostly non-developers and no-power users).

In my not-so-humble opinion, the majority of end users are simply wrong.
Even if they weren't wrong, I would use a batch file.

If I do end up implementing a CLI, then I'm happy if it helps some developer with an alternative reference to contribute a GUI in something else like gtk or Qt, but I am no good at GUIs written in anything that isn't GL so am not going to attempt one personally.

>And yes, C# and xbuild are both free and open source.

I don't know the details, but Stroustrup said it's proprietary and linked to a specific OS.
http://www.stroustrup.com/bs_faq.html#Csharp

>This approach would minimize the effort to port to new platforms after Linux (Mac OS X, even BSD).

I'm sure there are several C or C++ alternatives that work on all of those systems simultaneously.

An odd example, but if Clang or etc. had a compiler switch to recognize the Microsoft types like BYTE, WORD and DWORD, then I'm sure this would minimize the effort to port Project64 to other platforms; however I'm simply interested in actual portability and making the code more readable or even cleaner in cases.  As those type names assume a non-MIPS CPU and are misleading to have in the code anyway.

**Comment by aschwant** [comment]
I think a CLI would be a good checkpoint from which multiple cross-platform GUIs could be built. Having a nice API for setting the emulator state would make both possible and writing a CLI is easier and cleaner than building a GUI. Having a decent API would pretty much guarantee that someone will write a decent GUI to go with it. 

The core has windows dependent Structured Exception Handling (as long as the information in this thread is still relevant: http://forum.pj64-emu.com/archive/index.php/t-3688.html), which is a major issue to overcome before cross compatibility can be achieved. 

Makefiles suck. It is a real shame that make became as popular as it is. Personally I like the idea of scons, but it really doesn't matter that much. Build scripts are only for convenience and as long as the code does not depend on a specific compiler, changing the build script is relatively trivial, though annoying. 

I am a Linux programmer and would love to help out with making PJ64 cross platform, but I am in the middle of my last semester at my university and spare time is increasingly hard to come by at the moment.

**Comment by cxd4** [comment]
Well I am happy to hear the interest from everyone.  A Linux port I imagine would be one of the last things that would ever help zilmar profit off the users, but this, in the long run, is a much more socially effective strategy--both for teamwork and community in future gains.

>The core has windows dependent Structured Exception Handling (as long as the information in this thread is still relevant: http://forum.pj64-emu.com/archive/index.php/t-3688.html)

Yep, that is an unnecessary practice which a little R4300 experience can be used to replace.  The more direct and readable algorithm is to express address range boundary translation in the CPU.

A midway port of Project64 from MSVC to MinGW-gcc for Windows should be sufficient enough to break this particular barrier.  This was how I ported Azimer's Audio to Linux eventually.  (And MinGW is infamous for breaking several things--more than native GCC would--so this should help even more code compliance testing.)

>and writing a CLI is easier and cleaner than building a GUI.

Exactly this.  That's the only necessary thing for us to be concerned about prioritizing to overcome this long-term complex problem:  Simplicity, and in the future I am no less satisfied if it does help someone contribute a free GUI.  I would probably use it myself.

**Comment by JunielKatarn** [comment]
Not very relevant, but...

> it shouldn't be Clang-specific, it shouldn't be MSVC-specific, and it shouldn't be MSVC-or-Clang-exclusive

Agreed. I mention LLVM because of its (alleged) standards compliance and ABI compatibility with both GCC and MSVC's standard libraries. While code should become both standard and portable, I guess at some point @project64 might have to choose a compiler collection for "official" Project 64 cross platform releases. Personally, I would stick to that collection for my own development.

> Stroustrup said it's proprietary and linked to a specific OS

Well, he's wrong (as of today. His claim was true a few years back).
https://github.com/mono/mono
He is also wrong about Java:
http://openjdk.java.net/guide/repositories.html
Both language specifications are driven by a standards committee (not companies) and have open source implementations.

**Comment by darakian** [comment]
It'd be nice if we could sync with https://github.com/gonetz/GLideN64 as well.

**Comment by darakian** [comment]
With respect to the MS specific types is it possible to simple replace them with more portable types? ex. DWORD to uint32_t?

**Comment by project64** [comment]
> With respect to the MS specific types is it possible to simply replace them with more portable types? ex. DWORD to uint32_t?

this has slowly been happening

**Comment by darakian** [comment]
Is it as simple as just replacing keywords or is there a nuance in the code that I don't understand? I've been looking at the source and the lack of comments and odd function/variable names is making it very hard.

**Comment by cxd4** [comment]
It's what you said in most cases, but complex in others.

You can replace any instance freely, of BYTE with uint8_t, WORD with uint16_t, and QWORD with uint64_t.  The only real problem is `DWORD` vs. `uint32_t` because of an idiocy with the LLP64 ABI.

Other than that, we don't have to change all of them consistently--you can declare or pass calls to the function in one line with a BYTE, the other with a uint8_t.  They don't have to match right away everywhere.

>I've been looking at the source and the lack of comments and odd function/variable names is making it very hard.

Basically, if it's Windows-specific code (like GUI, recompiler interface, etc.) then leave it alone to disabled Windows types that won't be compiled for other platforms.  Otherwise if it looks like it's irrelevant to depending on any Windows OS memory/functions for management then it should be replaced with portable types.

Yeah some of the code is confusing like LW_VAddr(DWORD, DWORD "Value" (wut?), void*) because Value is generic and you don't know which needs to be a uint32_t, which needs to be a DWORD or a size_t.  But in most cases replacing with standard types is easy.

**Comment by darakian** [comment]
I'll give it a go later tonight or tomorrow. My main problem is that I don't have visual studio and I can't test a compile. Also one of the functions I found had something like
somefunction(bool thread){}
Why is a boolean called 'thread'!?

**Comment by cxd4** [comment]
Sorry to hear that.  I really should get to getting a MinGW script to compile Project64.  It will be like the free Visual Studio Express compiler downloads and have no UI to start emulation but at least let you test compilation without having to install 10+ GB of Microcrap.  (MinGW is between only 20 to 100 MB depending which components you install--C core is required, C++ Fortran Ada Java are optional.)

>Why is a boolean called 'thread'!?

No idea really; I mostly look more at the interpreter functions.

It might have been meant as an `int`, not a `bool` at some point, then it was changed to the Microsoft `BOOL` type (which really is just an int), then Lioncash contributed a PR to change most occurrences of "BOOL" to "bool".  Probably I just did as you did, got confused by it and put it off / never looked at Win32 threading stuff, just stuff that should make sense for all platforms.

**Comment by loganmc10** [comment]
@cxd4 Do you have a working version of project64 for Linux? I see that you are compiling static libraries, how are those used? Do you have a way to launch games from the commandline?

**Comment by cxd4** [comment]
@loganmc10 The static library model is inherent from the Microsoft Windows port of Project64.  I am just keeping that philosophy to reuse the same code while keeping it portable.  Most of the static libraries have no pertinence to Project64 itself but more to plugins or UI features.

Anyway, it runs now, but no X11 interface is established yet.  I will not be able to be worrying about that for a few weeks/months now when I am able to get back to this little project.

**Comment by Palatosino** [comment]
You think that this feature can evolve to be libretro compatible?

With this you can use any frontend that is libretro compatible like retroarch (very used in raspberry pi, because is used in RetroPie) or future RetroPlayer for Kodi (that will load libretro cores).

**Comment by TheEssemCraft** [comment]
Is using Qt to make a GUI an option?

**Comment by TheEssemCraft** [comment]
@Palatosino However, the Android version uses a mixture of Java and C++. I don't know if either is an option on the Pi.

**Comment by project64** [comment]
the java part for android is just UI. The core should be compatible on linux, just needs a UI to be written and probably a little bit of tweaking for the video plugin

**Comment by TheEssemCraft** [comment]
@project64 Ok! I'll try to write a UI using the original one as a base. Thanks!

**Comment by loganmc10** [comment]
@cxd4 i don't know much about Project64, but what about the RSP? I don't see that code in your static library scripts, and the RSP code doesn't look very portable

**Comment by project64** [comment]
@loganmc10 RSP plugin for lle is not portable and needs a lot of clean up. but there is a hle, one that does audio for android that can be used on linux

**Comment by cxd4** [comment]
@loganmc There is a MinGW script to build it on Windows without the static library model, but I don't remember why I stopped working on the Linux one.  Probably got tired of juggling across unmerged branches.  I certainly would not have ported the dynamic re-compiler code; that's probably the part you are seeing that doesn't look very portable.

A rather moot point.  Project64 is plugin-based, and RSP plugins in LLE exist for Linux already.

**Comment by TheEssemCraft** [comment]
So I guess that no one is gonna mention that MSBuild is available for Linux?

**Comment by cxd4** [comment]
Does it matter if it is?

**Comment by TheEssemCraft** [comment]
@cxd4 Well, MSBuild is what Visual Studio uses to build applications, so it might work if we build it for Linux using that.

**Comment by JunielKatarn** [comment]
@TheEssemCraft It does matter, but it's no silver bullet.
MSBuild works on major UNIX platforms (Mac OS, Linux). But only form .NET projects. Meaning, no native (C/C++) targets exist for those OSes.
I happened to start a project to create the appropriate mappings from MSBuild to Clang/LLVM on UNIX, but it's far from ready.

Finally, native MSBuild for UNIX is only part of the story. A large portion of the property sheets and VCXPROJ makefiles would need to be tuned to work seamlessly on non-Windows platforms.

**Comment by cxd4** [comment]
I have no idea what you two are talking about.

The project needs to have Windows-specific junk removed from it; that's all there is to it.  No makefiles, no property sheets, and no personal IDEs / MSBuild have absolutely anything to do with porting the RSP.  All of those things are related more to the preferential setup of the build environment and not the ability to get a build going for necessarily any platform at all.

Maybe you guys are talking about having some wrapper or whatever that could re-compile Windows-specific code into Linux-specific code?  But, that's hardly valid.  If that's considered porting Project64 to Linux then just install WINE to "port" it that way; it's easier.

**Comment by JunielKatarn** [comment]
> All of those things are related more to the preferential setup
Not what we mean at all, and modern, successful software native multi-platform projects leverage both clean code bases and unified make meta files for building.

I don't have time to educate you on the matter, but you can search on the internet about these revolutionary concepts.

**Comment by loganmc10** [comment]
You can't use MSBuild to build c/c++, so this whole thing is a pointless discussion.

To be honest it would probably be easier to work on mupen64plus which is already cross platform and stop supporting a guy that charges you to save games on Android...

**Comment by cxd4** [comment]
Exactly.  .NET and MSBuild are both entirely unrelated to the problem of C/C++.

They might have some sort of domain of bearing on MSVC or .NET.

>and modern, successful software native multi-platform projects leverage both clean code bases and unified make meta files for building.

I know, right?  It's called `make`.
https://www.gnu.org/software/make/

Or you could use CMake.
Or you could use SGI Make.

Or you could just write a shell script, which compiles inside Windows just fine under Git Bash.  But MSBuild is completely off the topic of what TheEssemCraft was asking about, which was porting Project64.

**Comment by oddMLan** [comment]
>To be honest it would probably be easier to work on mupen64plus which is already cross platform and stop supporting a guy that charges you to save games on Android...

To be fair mupen64plus is a collaborative team working on it whereas Project64 is doing most of the job of porting it by himself. For normal users, saves aren't a MUST to be able to play a game, so I don't have much of a problem with that being reserved to donors. People I know like to play Super Smash Bros in multi with their phones anyway most of the time. And you say, what about games that DO require saves like Zelda, etc. Consider also that the Android version is a WIP and still has a lot of room of improvement with compatibility and usability, so playing long game sessions may result in unstability. **Although maybe testers/contributors that help Project64 with code should get a code to get rid of the nag/limitations.** After all we will need those options the most to do more effective testing, like being able to test at specific points where games crash.  Donations solely drive the development towards it, allowing Zilmar to spend more time on this project. He has a family to take care of. I guess if people made more contributions to the Android version, the story would be different.

**Comment by cxd4** [comment]
Does the Windows build of Project64 require a code to activate it for builds you compile manually?

**Comment by LegendOfDragoon** [comment]
Project64 has a brighter future than m64p.

I think it would be cool to run Project64 on Linux. Then i'd have more incentive to use Linux more often and test out OpenGL on that.

Aside from UI and RSP plugin, what else needs to be done in order to run it on Linux?

**Comment by JunielKatarn** [comment]
It's a matter of removing Windows assumptions from the code base, and setting a bunch of makefiles (even some manual SH scripts would do).

It's a huge task, so it would be a good idea to go project by project.
I'd start with "common".

**Comment by loganmc10** [comment]
@LegendOfDragoon No project that nags you to pay up and charges you to save games has a bright future.

**Comment by project64** [comment]
@JunielKatarn Common should be fully able to compile on linux, same with glide64 and core ... 

Have a hle audio plugin for android that will work for linux as well .. there is UI, input and audio

**Comment by oddMLan** [comment]
>Does the Windows build of Project64 require a code to activate it for builds you compile manually?

The code for that has been in master for some time already so yeah, but the code is only for removing the nag when you start Project64 (much like WinRAR after the 'trial' period expires). Besides that there aren't any other limitations.

**Comment by oddMLan** [comment]
>@LegendOfDragoon No project that nags you to pay up and charges you to save games has a bright future.

If it bothers you that much you can remove the code for it and compile your own build, the project is open source may I remind you.
If there isn't someone out there compiling android apks with the activation code removed, you can be the first one to stick it to the man :smile:

**Comment by loganmc10** [comment]
Once it actually works I will, I've already come up with a name! pajama64, project64 for the lazy... I'd do the same for Windows if I didn't hate Visual Studio...

**Comment by oddMLan** [comment]
>No project that nags you to pay up and charges you to save games has a bright future.

I have one more point to add to this argument. I've seen quite a dozen of apps that rely on the same principle that Project64 does. Apps like SetCPU have an Play Store version that either is pay upfront or free but with ads/in app purchase to remove ads or free but with limitations/in app purchase to add functionality. However, you may download the apk from the official thread in XDA developers and have it completely for free, with just a nag to donate to keep the development active, but with no functionality limation. Each one of these approaches is a bigger compromise than the other. Personally I think Project64 took a good approach since if it was "pay upfront", the app would've had much less usage and limited testing in other devices (people like their free stuff). 

It's frequently seen that apps tend to adopt the freemium model to get better features, but being also able to enjoy of most of the features of the app without having to pay if one can't afford it or doesn't feel to. And they also tend to profit higher than the "pay upfront" ones. That's because you can test it and see if it meets your expectations, then pay for the rest of the features. If Project64 took the "pay upfront" model, people would leave extremely negative reviews and request a refund due to the fact that Project64 for Android is still a beta WIP. At least with the current model people can see if it runs slow or doesn't work with their games or if it has graphical bugs and then decide whether paying for the save feature is worth it. But knowing that Mupen64Plus AE is out there, why do you think people are still interested in Project64 on Android anyways :wink:

**Comment by loganmc10** [comment]
That's fine if the code is developed by the person collecting the money. Zilmar took the Android RSP code directly from mupen64plus (without any credit I might add, bsmiles called him out on that).

I think it's a dishonest way to go about things considering how many others contribute to the code base. He dint write Glide64 either for example

Once Project64 gets a decent Linux/Android port I'll just port it to RetroArch and people can use it for free that way, if it turns out to be better than mupen64plus

**Comment by project64** [comment]
@loganmc10 why so negative, you sound jealous or bitter or something. 

Project64 core is 99-95% mine, Yes I am using glide which I did not write, I am using bsmiles audio hle as well. I was cleaning up the code and try to make things cleaner, I was never being dishonest. When I changed the header and kept the license which is how GPL works, I copied the GPL2 headers I have been using that is like 10 lines instead of 100 as was there. I was meant to also copy the copyright line as well in the new version which I did not do in the clean up. Which was wrong a mistake not malicious and I fixed it as soon as it was pointed out to me. 

As I am using Glide64 and hle audio plugin that is what happens with GPL.

Is RetroArch GPL, I thought it was using a different license?

**Comment by loganmc10** [comment]
Lol don't you think if I was jealous I would be trying to make money myself? 

Anyway I guess it just bothers me to see a commercial product try to pawn itself off as some sort of open source collaborative effort. Your main motivation is money, it's the only reason an Android port was even attempted (since the calls to remove the adware/malware grew too loud I suppose). Anyway this is your project so I'll stop my soapboxing...

**Comment by project64** [comment]
@loganmc10 your free to have your own opinions, you can disagree with me as much as you like. I just want it to stay positive. 

Is money a factor yes, is it my only one no way. I would probably make more money working at McDonalds than I do working on project64 if you calculate the hourly rate. My ultimate goal would be to make money enough money where i can work on this full time  to make the best possible product, I know that will never happen, not even close. If I did not make the money I have I would have quite a long time ago like every other n64 emu author. 

My wife does not mind me working on this as the little extra helps give her gifts or just that little extra. 

I have never pawned project64 off as a some sort of open source collaborative effort. I made project64 open source for the benefit of the community not for my self.  I have always been open and helpful to any other emu writers,  Being able to use glide/bsmile hle audio is one of the few benefits from being open source. Yes other people do provide pull requests and fixes at times, but it is like 95-5 of the work I have put in to it.

**Comment by JunielKatarn** [comment]
> You mind giving some examples?

Mostly plugins (Audio, RSP), and possibly a non-platform-specific entry point (current main.cpp is Windows-specific).

A lot of code has been ifdef'd also, but there might be places that would require compiling to find out the missing gaps.

**Comment by JunielKatarn** [comment]
@project64 I have a couple of unrelated questions regarding recent updates to the core code base.
Is there a chat channel you use to discuss the project's internals? (I find the forums a little slow for this matter)

If there is none, you could set up a Gitter chatroom, which uses GitHub accounts and would be rather easy to connect with contributors:
https://gitter.im/home/explore#createcommunity

**Comment by project64** [comment]
I am often on irc in #project64 on freenode

**Comment by cxd4** [comment]
>The code for that has been in master for some time already so yeah, but the code is only for removing the nag when you start Project64 (much like WinRAR after the 'trial' period expires). Besides that there aren't any other limitations.

Well I have been kinda busy, so whatever is added to master as far as Windows-specific stuff is concerned is beyond my job to isolate.  This has to be my first day of even having enough time left in the day to get any wifi in almost a week.  And I don't have Windows installed anyway...been thinking about a VM of some kind though, maybe wine if I can get a feel for how it's supposed to install.

>Aside from UI and RSP plugin, what else needs to be done in order to run it on Linux?

"and RSP plugin"

I'm not honestly sure why you suppose different plugins have to do with porting Project64, RSP or otherwise.

The emulator just needs to be free of cruft from the Windows API or use a bunch of weird classes to continue using that with Macintosh APIs, X11 APIs and whatever else you want to redefine your own entire subsystem here in C++ for every operating system you port to.

>I think it would be cool to run Project64 on Linux. Then i'd have more incentive to use Linux more often and test out OpenGL on that.

It's way funner and more addictive to have an interest in OpenGL outside of emulation land.

And as valuable as I find the plugins specs, you do lose a lot of context management and application control with OpenGL when learning about the API just by toying around inside an emulator plugin.

I was never going to be successful with any OpenGL type of n64 thing until I learned how to use it for at least CAD and math research first.

>and setting a bunch of makefiles (even some manual SH scripts would do).

Makefiles and makefile configurations (and MS-centric advertising of things like MSBuild) have zero correlation with porting the code to Linux.  The make system of preference is simply an automated interface to how the build should be configured.

The only reason I add the SH scripts is for the convenience of testing the compile-time process outside Windows.  They could all be removed (or replaced with some other make system) and be portable all the same to whatever platform.

>@LegendOfDragoon No project that nags you to pay up and charges you to save games has a bright future.

Well, @loganmc10, supposing either that Project64 does or does not have a bright future, I see no dimension to compare for with respect to Mupen64Plus, which doesn't even have much of a present.  A future I suppose can exist without a present, but it's pretty worth noting that not much mainstream activity has ever existed in "Plus" ever since the project's debut with a contributed 64-bit recompiler and fully 64-bit-working port of the whole program.  It's mostly a free-for-all for various contributors, which is cool I think, and no mainstream continuance, and is otherwise dead.  Aside from GUI refinemenets it hasn't really improved on the original Mupen64 source base much at all; in fact a few at RetroArch have planned to just drop the core and use the original Mupen64 0.5 source base to fix Plus' regressions.

And since with little effort I fixed the 0.5 Mupen64 for building as 64-bit, I think the main difference in Plus is the title and the fact that it has a community watching it over IRC due to being the "new thing".  Although having a 64-bit dynarec is nice, too--but is unrelated to fixing N64 emulation, so much as making what's already broken in places even faster.

>Yes other people do provide pull requests and fixes at times, but it is like 95-5 of the work I have put in to it.

A good majority of open-source projects work that way @project64, where 95% of the work is still just 1 guy doing all that.  It happens to me, too, with every N64 thing I've done so far.  There are also plenty of exceptions that absolutely depend on a community of team open-source developers; unfortunately most people seem to incorrectly analogize this to video game land and piracy--two of the things that separate programmers from software engineers.
, 
>I will take a nagscreen any day, over having no gui..

GUIs are some of the most arbitrary-looking code ever.

They're full of added functions that seem to have absolutely nothing to do with what the goals people claim to be, really are.  At the end of the day whether you use a GUI, TUI, type in a command Unix-style, double-click a batch script or whatever, it's still going to be a N64 game running in a video window with rasterization for as many hours as you play it, so who honestly cares how fancy the Win32 GUI a few seconds before emulation starts looks?

That's one of the things that separates software engineers from programmers.  The latter are obsessed with what's attractive to audiences of human intellect, not what's pure, actual or efficient.

>To me, disregarding Windows users is far more off-putting than a nag screen.

Well, it stands to reason that it wouldn't be if you stopped using Windows, now, would it? :/

buttrly.  They do that for a number of reasons.

*Non-Plus Mupen64 didn't really disregard Windows users as much as Plus did, because it had forms of code-duplication for having the interface working under different OSes with different APIs.  Plus devs "simplified" this out of the code, knowing it would break the availability of some Windows builds.
*The POSIX standard, a central API through which all applications can achieve portability to most modern operating systems, is obeyed by all major operating systems except Microsoft Windows.  There's no real reason for that smarter than Microsoft trying deliberately to break ease of portability by making their OS as unique, broken and one-of-a-kind as possible to protect their profit methods.
*They're zealously anti-Microsoft, not always for fair reasons (like the above).  They're easily not the most evil entity in the world (especially considering how much control they have over the world to exercise that evil if it were worse), and as long as even Windows adheres to correct C standards I see no reason to drop support for it.

**Comment by LegendOfDragoon** [comment]
Speaking of copyright, since Project64 is copyrighted and some of its code got copied over to Mupen64plus, aren't they supposed to add Project64 to the copyright list in alist.c?

>It's way funner and more addictive to have an interest in OpenGL outside of emulation land.

Idk what I'd do with gfx APIs outside of emulation, other than make games.

>buttrly. They do that for a number of reasons.

I understand they wanted to "simplify" things, but I still don't like what they did. I honestly would rather just use Mupen64 0.5 over m64p.

>At the end of the day whether you use a GUI, TUI, type in a command Unix-style, double-click a batch script or whatever, it's still going to be a N64 game running in a video window with rasterization for as many hours as you play it, so who honestly cares how fancy the Win32 GUI a few seconds before emulation starts looks?

WEll, at first, i didn't even know how to use m64p. So it was pretty frustrating to setup and start a game. Now I got the hang of it but I still prefer having a fancy Win32 GUI. I feel that it's simply more convenient to me, than command line. Imo, having a ROM browser is nice, as opposed to having the drag the rom file into the cmd window lol. I like emulators that are convenient to setup / edit options.

**Comment by cxd4** [comment]
>Speaking of copyright, since Project64 is copyrighted and some of its code got copied over to Mupen64plus, aren't they supposed to add Project64 to the copyright list in alist.c?

IDK.  Depends what a copyright is, I guess.

I thought it was the right to copy something, or maybe that's copyleft.  Ask Stallman.

Although, with or without some GPL license, I think it's just good principle to acknowledge the copy pasta of code you never actually wrote.  I hardly want people to look at me thinking I wrote it, better or worse, based on code I just blindly pasted in.  You also make fewer enemies if you give credit whether you're obligated to or not, so I don't really see any disadvantage to just doing it and being happy with that out of the way.

>Idk what I'd do with gfx APIs outside of emulation, other than make games.

It's not particularly useful for emulation, either, unless you can think of a usage of OpenGL for emulation of a non-gaming system.  So really the basic reason you're left with is game design.

In which case, you'll love OpenGL 3.1 and up then, because that's exactly what I **don't** ever want to use it for.  Most people use it for game design, though, hence the premature deprecation in GL3+.

As far as what you could use GL for, if not gaming systems, well...um...anything related to what your monitor screen shows you?  Or time-sensitive things like movie players or video streamers.  Get that anime to stop lagging in crappy video players!

>I like emulators that are convenient to setup / edit options.

I sometimes prefer the GUI for editing options on a regular basis, but even then I still find myself going back and forth depending on the hour of the day and my mood, between that and just using simply the command line.

>WEll, at first, i didn't even know how to use m64p. So it was pretty frustrating to setup and start a game. Now I got the hang of it but I still prefer having a fancy Win32 GUI.

They redid the switch names, so I don't know how different it is from 0.5.  I never used their CLI.

>I feel that it's simply more convenient to me, than command line.

I type at about 95 words per minute.
I don't usually notice a firm difference between GUI and good CLIs.

Except maybe on laptops, because I wish I could rip this stupid touchpad right out.  Just typing seems to be enough to hover my palms closely over it enough so the mouse cursor moves way over to the opposite end of the screen away from where I went.  Stupid-ass mouse.  Why does Microsoft keep inventing animals that I can't ever catch and eat?

>Imo, having a ROM browser is nice, as opposed to having the drag the rom file into the cmd window lol.

I highly doubt that's what they're requiring you to do, or even the method they expect people to use.

Proper usage of a command-line utility requires RTM, or read the source to learn the options.

Also, be somewhat inventive at times.  If the commands are too long to type out in a timely and effortless manner, have some batch files in a directory.  There's your ROM browser right there:  Windows Explorer.  It's almost like being able to double-click the ROM files themselves to open them directly except it's way more customizable (i.e., scriptable).

>I understand they wanted to "simplify" things, but I still don't like what they did. I honestly would rather just use Mupen64 0.5 over m64p.

So would I.  Just not for that reason.

But as far as "what they did", you would probably know more about that than I would because save for the RetroArch core I've never actually used Mupen64Plus seriously for anything, short of testing my plugins for forward-compatibility with their broken model.

In principle, it was good to separate the GUI into a separate program from the actual emulator, because the emulator doesn't require a GUI to emulate, and the GUI always requires platform-specific code.

Still, even with Mupen64 0.5, I often go back and forth between the GTK vs. the no-GUI Mupen64 (the `mupen64_nogui` command that the official Windows release doesn't pre-include).  The latter is actually sometimes easier than the GUI version.  Even if I have to write out:
```
./mupen64_nogui --fullscreen -g plugins/z64gl.so -a plugins/audio.so -i plugins/keyboard.so -r plugins/rsp.so --emumode 1 /n64/NSME-001.n64
```
or whatever, a simple press of the up-arrow key will read that in from the cache faster than loading the GUI.  So I could open the GUI m64 and use the mouse to double-click the ROM after scrolling through the browser, or I could just hit UP ARROW then ENTER both in less than a tenth of a second and play the thing already.  GUIs are so inefficient.

And don't forget about <TAB> completion.  Most of that command I just gave you wasn't typed letter-for-letter.

**Comment by LegendOfDragoon** [comment]
>IDK. Depends what a copyright is, I guess.

I don't get why the same people who complained about zilmar for not including the copyright, would take Project64's code and not include Project64 in the copyright.

>Although, with or without some GPL license, I think it's just good principle to acknowledge the copy pasta of code you never actually wrote.

Agreed

>So would I. Just not for that reason.

There are several reasons, such as modern versions of M64p having their own spec (which didn't even bother to make improvements such as keeping track of RDRAM size).

>I highly doubt that's what they're requiring you to do, or even the method they expect people to use.

Fair point lol. Can't say I'm too great with using CMD apps, but i was pretty bad in my noob days :smile: . Guess I should be more acquainted with stuff like batch scripting.

>In which case, you'll love OpenGL 3.1 and up then, because that's exactly what I don't ever want to use it for. Most people use it for game design, though, hence the premature deprecation in GL3+.

I probably would like OpenGL 3.1 and up. My dilemma is that i want to support older hardware. I may learn it someday anyway, since my desktop fortunately supports it.

>Or time-sensitive things like movie players or video streamers. Get that anime to stop lagging in crappy video players!

I would be more inclined to do something like that, if I felt I could implement something significant that is lacking in other programs. Can't see myself making a media player that can even replace current ones.

>GUIs are so inefficient.

Even though clicking can be slower, I still generally prefer clicking over typing :smile: . Not sure why tbh. When I use 1964 I actually click the pause button to pause instead of pressing F3 and I actually do the same on PJ64 as well.

**Comment by cxd4** [comment]
>I don't get why the same people who complained about zilmar for not including the copyright, would take Project64's code and not include Project64 in the copyright.

Do you propose that bsmiles used zilmar's code without including him in the copyright?

Because I don't know.  I don't follow what they do.

I know most of the rest of the Mupen64Plus team did this with the spec files.  The copyright was cited to people who had never touched the plugin specs (like Hacktarux or Richard42) any more than zilmar had done but was not cited to him.  But that itself I'm pretty sure wasn't the doing of bsmiles.

>Agreed

For anything shallow of popularity to open-source contributors, like emulation projects, the choice of software license hardly seems to matter as far as forbidding certain things or making rules against doing things.  The only purpose the choice of license for such projects would have is just what is allowed to encourage people who care about the license to want to contribute, but what's banned or forbidden rarely anybody in video game emulation land is going to view the same way as with more popular and developer-supported projects.

>I probably would like OpenGL 3.1 and up. My dilemma is that i want to support older hardware. I may learn it someday anyway, since my desktop fortunately supports it.

You would just assign the function pointers to call subroutines using OpenGL 1.1 or 1.0 functions for where the presence of GL 3.1 or 2.0 extensions is found to be incomplete.

>I would be more inclined to do something like that, if I felt I could implement something significant that is lacking in other programs. Can't see myself making a media player that can even replace current ones.

It's usually pretty easy to find optimizations in others' code.  The people making and/or selling these large, immensely popular video players are hardly concerned with being perfectionist at micro-optimizations because time is money.

>Even though clicking can be slower, I still generally prefer clicking over typing :smile: . Not sure why tbh. When I use 1964 I actually click the pause button to pause instead of pressing F3 and I actually do the same on PJ64 as well.

Um well with 1964 I usually click the pause button since the whole UI gets buggy a lot of the time anyway.  Just hitting F2 (or was it F3) to pause can freeze the whole UI if I'm on the wrong system.  Anyway I was no longer able to get 1964 to work one single bit on my last Windows desktop so I barely remember anything about it.

But anything else it's way more instantaneous to just hit F2 to pause/resume it than to reach for a mouse that's just going to keep running away from me no matter how far I chase it.

**Comment by LegendOfDragoon** [comment]
>Do you propose that bsmiles used zilmar's code without including him in the copyright?

Logan took some commits from PJ64's version of the HLE RSP and ported it to M64p's repo without adding PJ64 to the copyright.

>I know most of the rest of the Mupen64Plus team did this with the spec files. The copyright was cited to people who had never touched the plugin specs (like Hacktarux or Richard42) any more than zilmar had done but was not cited to him. But that itself I'm pretty sure wasn't the doing of bsmiles.

That's honestly pretty sad, considering only zilmar got flak for not including copyright.

>It's usually pretty easy to find optimizations in others' code. The people making and/or selling these large, immensely popular video players are hardly concerned with being perfectionist at micro-optimizations because time is money.

That's a good point. Maybe I should look into finding some optimizations some day for stuff like MPC.

>But anything else it's way more instantaneous to just hit F2 to pause/resume it than to reach for a mouse that's just going to keep running away from me no matter how far I chase it.

Part of my reason is that i hate having to hold Fn before pressing F-keys. I don't pause too often so I just click on the pause button.

**Comment by loganmc10** [comment]
@LegendOfDragoon my issue is people profiting/selling a product based on the work of others. I think the whole copyright thing in open source software is rather dumb and just serves to limit the openness of the software, I don't add my copyright to any of the code that I write.

**Comment by cxd4** [comment]
That and copyright is one of those things that supposedly self-applies itself in our legal system.

So the only way to not have copyright on code is to copyright to some not-me guy or use licenses that your government gets along with saying no copyright exists.

>Part of my reason is that i hate having to hold Fn before pressing F-keys. I don't pause too often so I just click on the pause button.

Well WTF then you need to get a better F'n keyboard then!

I've never heard of a keyboard/laptop that won't let you hit F2 straight away.  I do have to hold the Fn key for Insert --> PrintScreen though, but that's only if I want to use KDE to take screenshots instead of Xfce anyway.

Even on those public Windows 10 computers (which I hate seeing every day), I use F2 in Windows Explorer to rename files without moving the mouse or clicking and Alt+F4 all the time to close windows (as well as shut down / log off because if you have no windows open at all then Alt+F4 does that).

You need to get a keyboard that doesn't force you to depend on masochistic inefficient contraptions like that mouse animal MS started producing.  It will save you many rolls of TP.

Your other reason is probably because you don't use a keyboard to play N64 games like I do; maybe you have a real N64 controller and some kind of adapter.  Maybe it doesn't matter which you choose then in that case but chasing that mouse is still redundant.

**Comment by LegendOfDragoon** [comment]
>@LegendOfDragoon my issue is people profiting/selling a product based on the work of others.

In Project64's case, zilmar simply accepts donations. I don't consider it selling, because it's open source. You can say he's profiting, but I don't think these 3rd party plugins are the reason people are donating to project64. The emulator core itself, is what made Project64 popular. It's still one of few emulators that can even run Gauntlet Legends in LLE and some other games.

>Well WTF then you need to get a better F'n keyboard then!

I think it's just the default settings for this Dell laptop. I'm sure there's some way to disable it, but I just never bothered lol. I really should look into it, now that I think about it.

>Your other reason is probably because you don't use a keyboard to play N64 games like I do; maybe you have a real N64 controller and some kind of adapter. Maybe it doesn't matter which you choose then in that case but chasing that mouse is still redundant.

I usually use keyboard since my gaming is usually just casual. I do want to start using a controller again. Perhaps during winter break :smile: , while I start playing some RPGs. Another issue is that the F-keys are small on this laptop keyboard, sometimes I press the wrong key by accident. When I play on desktop, i use the F-keys because they are bigger and i don't have to hold Fn.

**Comment by cxd4** [comment]
>I usually use keyboard since my gaming is usually just casual.

Well I don't use keyboards to be casual.

I use it because that's why I pwnz0r @ExtremeDude2 every time we played _Super Smash Bros._ over NetPlay and I'd win because I play with a keyboard while he has to use his dumb Xbox controller.  O_O OR EVEN WORSE A REAL N64 CONTROLLER LOL WUT A NUBLET!

>You can say he's profiting, but I don't think these 3rd party plugins are the reason people are donating to project64. The emulator core itself, is what made Project64 popular.

Easily disagreeable.

When it comes to entertainment, people go by what they see, hear, and touch, not CPU internals.

There were no plugins more simple yet excellent than Jabo's for many years.  Your measure of what makes Project64 popular (for example, does it have a better internal CPU to get more games to boot like _Gauntlet Legends_ with a hack that could be added to any emulator) is more case-by-case with some games--popularity implies majority, and the majority of human beings are idiots man let's face the fact?  So I agree with your standard of why Project64 should be considered more popular, but why it **is** in actuality more popular to people in majority?  liek srsly m8

>In Project64's case, zilmar simply accepts donations.

Nah, I think you're getting into matter-of-perspective territory there.
http://www.dictionary.com/browse/donation?s=t

He *names* them donations.  The reason why the money was sent (charity, good will, purchase, donation, wanting something in return) is determined by the person sending the money, not the person accepting the money.

You're not allowed to make money off Nintendo's games, so the only way to really get money from Project64 without as much risk for being sued is to say that it is a donation regardless of whether it is in fact one.  You're citing what's the term legally, not lawfully.

Here's what I call neighbors giving away the treasure that is their trash (clothes that won't fit them, food they can't eat anymore, MarathonMan's laptop, theboy181's tester "loyalties", whatever...) _and then getting mad at you for not returning the hidden, implied favor_ or using their trash-into-treasure gift in the way they secretly intended you to use it:  **Investing.**

See, if they specifically tell you that you'll end up giving them back something years later in return for it, or if they specifically tell you that you need to use the "gift" according to their rules, how much you call this a "gift" starts to change in your mind based on what they say.  Or, they say nothing, _but that doesn't mean they think nothing!_  You don't have to **say** the rules you are thinking in your mind when giving somebody a hidden bomb while calling it a gift; doesn't make it a gift.

A gift implies that it's given to you out of the generosity of their hearts; what they're doing is an **investment** (usually a very shallowly conceived and foolish one, but some people are smart) that is fixed to strategically benefit them in the long run.  Many people "give" things because they overthink good karma or think it will get them into heaven or whatever, not out of generosity and not out of love.  It's not a gift; it's just self-deception and expecting something worthwhile of material value in return.

Those are the kinds of people who "give" so that they, in their minds and egos, accumulate power and things to hold over people, if not to actually say it outright that they're owed a favor in return.  Giving isn't about receiving; it's about love.  And let's face it:  Most people just **hate** whatever pop-up like thing pj64 is doing nowadays.  Doesn't mean no donations are happening but doesn't mean every single send of money is a donation just because that's what you're seeing zilmar call it.

>I don't consider it selling, because it's open source.

How easy it is to not buy it (e.g. compile the open-source project) does not dictate whether it is considered buying or donating.  The reason the offer exists and is always accepted for does.

The only thing that being open-source matters is how dumb the people are for buying it instead of compiling it, if the money they gave wasn't meant as some sort of selfless donation or, again, a gift.

**Comment by LegendOfDragoon** [comment]
>Well I don't use keyboards to be casual.

I'm fine with using keyboard for fighting and racing games. I just wouldn't want to use it for RPGs.

>When it comes to entertainment, people go by what they see, hear, and touch, not CPU internals.

That is indeed true. When I first started playing n64 on emulator, i chose PJ64 because it had the highest ratings (around the time 1.6 came out). I'm sure end users appreciate its UI a lot more than m64p though :smile: .

>You're not allowed to make money off Nintendo's games, so the only way to really get money from Project64 without as much risk for being sued is to say that it is a donation regardless of whether it is in fact one.

I get that, but the reason i still consider it a donation is because you can get the same product for free if you do not wish to pay. You can either compile it, or get the builds directly from zilmar just by asking (if I understand correctly).

>The reason why the money was sent (charity, good will, purchase, donation, wanting something in return) is determined by the person sending the money, not the person accepting the money.

That's a very good point. Can't say I am confident in knowing what most of their reasons for donating are, but I'm optimistic in that regard :smile: .

>and then getting mad at you for not returning the hidden, implied favor or using their trash-into-treasure gift in the way they secretly intended you to use it: Investing.

Wow I can definitely relate to that. I'm sure they wouldn't have been disappointed had they been honest about it in the first place :smile: .

>The only thing that being open-source matters is how dumb the people are for buying it instead of compiling it, if the money they gave wasn't meant as some sort of selfless donation or, again, a gift.

I'm having a very hard time believing that the majority of recent PJ64 donors are that silly. I could very well be wrong, but I really hope I'm not. I'm aware that some donors may indeed be that silly, but to think that the majority might possibly be that way, is very puzzling.

**Comment by cxd4** [comment]
>Wow I can definitely relate to that. I'm sure they wouldn't have been disappointed had they been honest about it in the first place :smile: .

I see it almost every day.  I call it investing, not giving.

Many people giving gifts in the public world are either secretly needy in some arena themselves or just trying to reward their own egos.  Some people seem to think it necessarily repents for their wrongdoings.

For the most part, it's hard for me to give a genuine gift, when I do do it.  It involves respect, love, and sacrifice (sometimes self-sacrifice).  It's not something I usually feel is easy to do without partially feeling for the person I'm helping out for some reason.  Thinking that maybe karma will repay you someday for it as well is fine, but looking for people to give things to for just that reason is stupid and shallow-minded.

I've seen clear cases of people "investing" in other people by giving them something under the guise that it's a gift, then getting upset at it not resulting in what they wanted it to result in.  Most gifts are strategic investments in using other people.  Some say it's better to give than to receive:  That's because the person giving is naturally (for the reason of being able to give in the first place) more equipped, resourceful or powerful than the person needing to receive, and is usually seeking out even more power to hold over/against people by giving them gifts.  (And the more power most men have, the more ambitious they are about seeking more.  It's similar with money.)

>I'm fine with using keyboard for fighting and racing games. I just wouldn't want to use it for RPGs.

Doesn't the N64 only have like, 5 RPGs available for it?  Unless technically we should count Zelda.

Stuff like GoldenEye I don't do too well with a keyboard as I did with a real N64 controller.  I'd like to buy an adapter to use a non-keyboard controller someday mostly for the purpose of being able to lean all the way back in my chair instead of having to sit all the way up to the keyboard.

>That is indeed true. When I first started playing n64 on emulator, i chose PJ64 because it had the highest ratings (around the time 1.6 came out). I'm sure end users appreciate its UI a lot more than m64p though.

I first started playing when someone copied Nemu64 off their flash drive to mine.

My next home PC was an old 90s machine with barely a 32-bit operating system on it and no accelerated graphics card, so barely any emulators/plugins worked at all.

And yeah I think subconsciously most users liked Project64 the most because of its elegance and UI without even realizing that was their own reason, but nowadays I'd guess it's more because it's active, developed, and the only emulator being progressed at all.

>I get that, but the reason i still consider it a donation is because you can get the same product for free if you do not wish to pay. You can either compile it, or get the builds directly from zilmar just by asking (if I understand correctly).

Technically everything's for free, if we're optimistic.

Also, nothing's for free, if we're pessimistic.  It's a real economic fact that nothing's for free.

Having to walk 5 miles to get to work is beneficial to me physically but even less "free" than paying money for something you can perfectly well afford.  So in the context you put "free" in, it's relative.

Some people would rather walk 8 miles to get a "free" build of Project64, than read how to compile.

Dunno if you find that hard to believe, but I'm pretty sure I've witnessed it in their attitudes.

So maybe you're right in some ways but it's still perspective.  The person giving the money in their mind decides whether it's free, a gift, a donation, etc. or not.

>I'm having a very hard time believing that the majority of recent PJ64 donors are that silly. I could very well be wrong, but I really hope I'm not. I'm aware that some donors may indeed be that silly, but to think that the majority might possibly be that way, is very puzzling.

How silly?  Silly enough to purchase Project64 instead of reading how to compile it for free?

Well I'm happy you feel that way.  Not saying I agree at all necessarily, but it sounds like you would do well in the Linux world.

Also, I wouldn't even call it "silly" necessarily.  Money means different values to you, me, and whoever donates/buys.  Why would somebody working full time and getting good money take time out of their personal or work live to read how to compile, when they could just buy a code to activate Project64 for $20 instantly?

You may call it foolish (and **for me** to do I think you are right), but time is sometimes more valuable than money.  And the RDB tweak to deactivate the donation nag also takes time.  Again, easy for you and I to do, not easy for someone trained to work all their life with non-computer things.  If they had their heads on forwards it's incredibly easy, but convincing them of that and teaching them is sometimes more time-consuming than for them to just pay zilmar $20 or whatever.  It's true.

**Comment by TheEssemCraft** [comment]
What are the command line arguments for Project64?

**Comment by cxd4** [comment]
```
git clone https://github.com/project64/project64.git
cd project64
git checkout -b merge-all-the-PRs
git pull https://github.com/cxd4/project64.git
git checkout master
git merge merge-all-the-PRs
```

**Comment by TheEssemCraft** [comment]
@cxd4 And what did you add? How do I compile it?

**Comment by JunielKatarn** [comment]
> And what did you add? How do I compile it?

@TheEssemCraft You can follow the instructions [here](http://pj64wiki.com/index.php?title=Building).

**Comment by TheEssemCraft** [comment]
@JunielKatarn None of the required programs to build are on Linux.

**Comment by cxd4** [comment]
No, that's just the resident Microsoft advertisements bot.  If he were a live human he would see this topic wasn't about MSBuild autism, but porting to Linux.

**Comment by JunielKatarn** [comment]
> None of the required programs to build are on Linux.

@TheEssemCraft Oops, my bad. I was going through unread threads and didn't notice this was the Linux one.

No way to make the whole thing in UNIX yet, except Android.

**Comment by cxd4** [comment]
>@cxd4 And what did you add? How do I compile it?

You make it by executing the `make` shell scripts.

That compiles and links it for all variants of Unix, except Android.  (This may or may not have changed ever since zilmar's present commits...if it happens that he broke anything let me know; I will check and send another PR.)

**Comment by cxd4** [comment]
https://github.com/project64/project64/tree/master/Source/Script/Unix

I was initially hesitant to answer your question because this is not something we were supposed to document at this time.  The Linux version builds just fine but has no interface to start the interpreter.  I didn't answer your question earlier because the necessary communications for tending to this matter have stayed on hold.

Also, I now realize the `project64.sh` (needed for building the UI) is absent from that folder...hm, now that I think about it I must have kept the file on-disk for me only until the other PRs to get the UI working were merged.  I forgot this never happened. :) Guess that's why you only have the `project64-core.sh` to compile the core with, then.

**Comment by TheEssemCraft** [comment]
@cxd4 Thanks for the info! :) I'll try to build it when I get home. Maybe I can redo the UI in Qt so that it has better cross-platform support?

**Comment by cxd4** [comment]
This would be easier if a UI existed at all for non-Android Linux platforms.

I had started a command-line interface to using Project64 with command-line arguments.  (The more I think about it, the more I remember none of those should be required, since the Project64/Config/ files pretty much have everything set up.)  I was unable to continue in good conscience with all the distractions of unported things, like fetching the full path to the current working directory on POSIX-compliant systems.  I had written a tried and true method of doing this in the C++ classes that would get the full current directory path on both Android as well as Linux, but the changes were never merged, I moved, I replaced my life, I got tired of waiting, blah blah blah.

**Comment by cxd4** [comment]
Oh, but to answer your question, you are welcome to contribute one if you find the time.

**Comment by xxmadbrinexx** [comment]
I'm not sure if I'm necroposting, but I'm having trouble running Project64 in wine. It kept saying DirectX failed to initialize, so I installed DirectX on wine. But it continued to fail... Going to plugin settings doesn't help, since it just gives me the same error. Can anyone help?

**Comment by cxd4** [comment]
Didn't think you'd have needed to.  Direct3D plugins I bundled with Nemu64 work well under WINE.

That is to say, without me having any knowledge of installing DirectX to WINE, as I've never done that.

Independent of those matters, there is less performance latency from API translations if you use a Windows OpenGL plugin under WINE as opposed to a Direct3D one, or you can try a different Direct3D plugin with less requirements.  If you're running Project64 2.something in WINE then you may have to edit the Config/Project64.cfg file in a text editor to change the video DLL since its settings forms GUI isn't supported by WINE last I checked.

**Comment by TheEssemCraft** [comment]
@xxmadbrinexx Have you tried using `winetricks`?

**Comment by Jj0YzL5nvJ** [comment]
Other option is use [Wine Staging](https://wine-staging.com/).

**Comment by xxmadbrinexx** [comment]
yes, that's what I used to install directx

**Comment by xxmadbrinexx** [comment]
@Jj0YzL5nvJ I'll try that.

**Comment by xxmadbrinexx** [comment]
@cxd4 I'll try if I can.

**Comment by oddMLan** [comment]
@xxmadbrinexx Your avatar is triggering.

**Comment by cxd4** [comment]
Actually easiest may be to just rename an OpenGL plugin to the filename used by Jabo's Direct3D (assuming Project64 still has that as the default video plugin because I wouldn't know).

So go to $project64/Plugin/GFX/ and there probably will be two DLL files, one named PJGlide something and the other Jabo_Direct3D or something similar to that.  If you just use the PJGlide one that won't require installing DirectX to WINE, or you can download a different (probably better) OpenGL plugin.

So select the Jabo DLL, F2 in Windows Explorer, Ctrl+C, Enter/Esc, delete the file or rename it, select the PJGlide DLL, F2, Ctrl+V to copy the file name over to that one so PJ64 on WINE uses that without having to go through the broken GUI config to set it.

**Comment by xxmadbrinexx** [comment]
Alright, thanks

**Comment by codeman101** [comment]
I kind of skimmed through this thread but I don't understand how it turned into a conversation about running the app in WINE. Most of this thread makes it sound like it's possible to port the windows specific code over to Unix. Is that still in progress?

**Comment by project64** [comment]
@codeman101 there should be no windows specific code in the core at the moment, it should really just require some one building a ui and linking everything together

**Comment by project64** [comment]
closing old issue, if this is still an issue it can be reopened

**Comment by Mhowser** [comment]
I'm In favor of keeping it open until a Linux version gets released. Has there been any progress towards that?

**Comment by project64** [comment]
I had some one mention on discord they were looking a bit at it, but there is no point keeping a issue open for it. 

I want the issues to be some actionable, otherwise you end up with a list of items that it is hard to find if there are things that actually require work or not. 

We need a place for features that would be liked, the issue tracker I do not think is the right place for it, tho I do not know where the right place is.

**Comment by sonofachamp** [comment]
You could try the [projects board](https://github.com/project64/project64/projects) to track non-issue related items.
